### PR TITLE
Feature[MCMPCSP2-968]: AutoDiscover and Generate API Docs

### PR DESCRIPTION
**Pull request by rigdomin-cisco**
**Note**: Will squash after final feedback

- [x]  AutoDiscover Services & Deployments
- [x]  K8s events trigger process to fetch API specs
- [x]  Fetch API specs by calling **/swagger.json** API
- [x]  Process returned API specs
- [x]  Provide processed API specs to Dapperdox documentation generator

Procedure to process API specs after they are fetched from their respective services:

1. Unmarshal API spec from **/swagger.json** API response
2. Remove Private Paths, Methods, Definitions
3. Apply grouping extension
4. Apply rewrite specifications
5. Marshal processed API spec

### CODE REVIEW

File: discover/apispecs.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+	wraperrors "github.com/pkg/errors"
+	"github.com/spf13/viper"
+
+	"github.com/kenjones-cisco/dapperdox/config"
+)
+
+const (
+	visibilityExt = "x-visibility"
+
+	groupByPrivate = "Private Cloud Services"
```
Comment (kenjones-cisco) [inline discover/apispecs.go]: for the groupBy* types, these should be done via configuration otherwise this would be specific to our internal use.

File: discover/apispecs.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+	groupByPublic  = "Public Cloud Services"
+	groupByCore    = "Platform APIs"
+
+	// key values swagger spec extensions.
+	extKeyMCMPComponentType = "x-mcmp-component-type"
```
Comment (kenjones-cisco) [inline discover/apispecs.go]: need to do this generically to avoid it being specific to our internal use.

File: discover/apispecs.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	extKeyGroupBy           = "x-groupby"
+)
+
+func (d *Discoverer) fetchAPISpecs() map[string][]byte {
+	// if there are rewrite configuration defined,
+	// ensure we are able to load them
+	var rewritesDoc *loads.Document
+
+	if viper.GetString(config.SpecRewrites) != "" {
```
Comment (kenjones-cisco) [inline discover/apispecs.go]: What is the purpose of this configuration? How would someone make use of this?

File: discover/controller.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+		}
+
+		return nil
+	})
+
+	return nil
+}
+
+// AppendDeploymentHandler notifies about change to the deployment catalog.
+func (c *catalog) AppendDeploymentHandler(f func(*models.Deployment, models.Event)) error {
```
Comment (kenjones-cisco) [inline discover/controller.go]: why is Deployment now being used instead of Instances?

File: discover/controller.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
+// Copyright 2017 Istio Authors
```
Comment (kenjones-cisco) [inline discover/controller.go]: given all of the previous Istio logic has been removed, this is very generic k8s now, this could be removed.

File: discover/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
+// Copyright 2017 Istio Authors
```
Comment (kenjones-cisco) [inline discover/controller_test.go]: remove this, see comment on controller.go

File: discover/discover.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/kenjones-cisco/dapperdox/discover/models"
+)
+
+// Discoverer represents the state of the discovery mechanism.
+type Discoverer struct {
+	services watcher
+
+	data   *state
+	stop   chan struct{}
+	router *mux.Router
```
Comment (kenjones-cisco) [inline discover/discover.go]: what is the router needed for?

File: discover/discover.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+	case models.EventAdd, models.EventUpdate:
+		d.data.deployments.Insert(dpl)
+	case models.EventDelete:
+		d.data.deployments.Delete(dpl)
+	}
+
+	d.discover()
+}
+
+func inIgnoreList(name string) bool {
```
Comment (kenjones-cisco) [inline discover/discover.go]: this is not needed, can always use `sets` from https://github.com/kubernetes/apimachinery/tree/master/pkg/util/sets so that we do not need the package variable and can avoid the need for a loop as well.

File: discover/discover.go [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	// copy Viper configurations to avoid any loss of viper
+	// due to high number of goroutines in Auto-Discover process
+	vprcpy := viper.GetViper()
+
+	ctlg := newCatalog(client, catalogOptions{
+		WatchedNamespace: vprcpy.GetString(config.DiscoveryNamespace),
+		ResyncPeriod:     vprcpy.GetDuration(config.DiscoveryInterval),
+		DomainSuffix:     "cluster.local",
```
Comment (kenjones-cisco) [inline discover/discover.go]: lets make this configurable

File: discover/models/event.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
+// Copyright 2017 Istio Authors
```
Comment (kenjones-cisco) [inline discover/models/event.go]: remove

File: discover/models/event_test.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2017 Istio Authors
```
Comment (kenjones-cisco) [inline discover/models/event_test.go]: removed

File: discover/models/service.go [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+	// Note that HTTP/1.0 or earlier may not be supported by the proxy.
+	ProtocolHTTP Protocol = "HTTP"
+	// ProtocolTCP declares the the port uses TCP.
+	// This is the default protocol for a service port.
+	ProtocolTCP Protocol = "TCP"
+	// ProtocolUDP declares that the port uses UDP.
+	// Note that UDP protocol is not currently supported by the proxy.
+	ProtocolUDP Protocol = "UDP"
+	// ProtocolMongo declares that the port carries mongoDB traffic.
+	ProtocolMongo Protocol = "Mongo"
```
Comment (kenjones-cisco) [inline discover/models/service.go]: drop this

File: discover/models/service.go [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+	// ProtocolRedis declares that the port carries redis traffic.
+	ProtocolRedis Protocol = "Redis"
```
Comment (kenjones-cisco) [inline discover/models/service.go]: drop this

File: discover/models/service.go [hunk 14]
```diff
@@ -0,0 +1,192 @@
+// Copyright 2017 Istio Authors
```
Comment (kenjones-cisco) [inline discover/models/service.go]: remove

File: discover/models/service_test.go [hunk 15]
```diff
@@ -0,0 +1,231 @@
+// Copyright 2017 Istio Authors
```
Comment (kenjones-cisco) [inline discover/models/service_test.go]: remove

File: discover/queue.go [hunk 16]
```diff
@@ -0,0 +1,141 @@
+// Copyright 2017 Istio Authors
```
Comment (kenjones-cisco) [inline discover/queue.go]: remove

File: discover/queue_test.go [hunk 17]
```diff
@@ -0,0 +1,89 @@
+// Copyright 2017 Istio Authors
```
Comment (kenjones-cisco) [inline discover/queue_test.go]: remove

File: discover/router.go [hunk 18]
```diff
@@ -0,0 +1,94 @@
+package discover
```
Comment (kenjones-cisco) [inline discover/router.go]: comment provided offline to avoid re-building everything that already exists

File: fixtures/rewrites.yaml [hunk 19]
```diff
@@ -0,0 +1,24 @@
+securityDefinitions:
+  token:
+    type: oauth2
+    tokenUrl: https://cloudsso.cisco.com/as/token.oauth2
```
Comment (kenjones-cisco) [inline fixtures/rewrites.yaml]: use examples.com here

File: go.mod [hunk 20]
```diff
@@ -28,4 +31,8 @@ require (
 	github.com/spf13/viper v1.7.1
 	github.com/unrolled/render v1.0.1
 	golang.org/x/net v0.0.0-20201202161906-c7110b5ffcbb
+	gopkg.in/inf.v0 v0.9.1 // indirect
+	k8s.io/api v0.0.0-20180308224125-73d903622b73
+	k8s.io/apimachinery v0.0.0-20180228050457-302974c03f7e
+	k8s.io/client-go v7.0.0+incompatible
```
Comment (kenjones-cisco) [inline go.mod]:
```
	k8s.io/api v0.0.0-20180628040859-072894a440bd
	k8s.io/apimachinery v0.0.0-20180621070125-103fd098999d
	k8s.io/client-go v8.0.0+incompatible
```

File: main.go [hunk 21]
```diff
@@ -54,12 +55,19 @@ func main() {
 
 	log.SetLevel(viper.GetString(config.LogLevel))
 
+	// start auto-discovery process
+	d, err := discover.NewDiscoverer()
```
Comment (kenjones-cisco) [inline main.go]: needs to be an optional feature

File: .golangci.yml [hunk 22]
```diff
@@ -379,6 +382,21 @@ issues:
         # - unparam
         # - wsl
 
+    - path: config/configurations.go|main.go
+      text: "forbidden by pattern"
+      linters:
+        - forbidigo
+
+    - path: formatter/markdown.go|handlers/reference/reference.go
```
Comment (kenjones-cisco) [inline .golangci.yml:397]: add a TODO, as this should be fixed

File: discover/controller.go (hunk 5 above)
Comment (rigdomin-cisco) [reply discover/controller.go]: done

File: discover/controller_test.go (hunk 6 above)
Comment (rigdomin-cisco) [reply discover/controller_test.go]: done

File: discover/discover.go (hunk 7 above)
Comment (rigdomin-cisco) [reply discover/discover.go]: removed

File: discover/discover.go (hunk 8 above)
Comment (rigdomin-cisco) [reply discover/discover.go]: simplified

File: discover/discover.go (hunk 9 above)
Comment (rigdomin-cisco) [reply discover/discover.go]: suffix configurable

File: discover/models/event.go (hunk 10 above)
Comment (rigdomin-cisco) [reply discover/models/event.go]: done

File: discover/models/event_test.go (hunk 11 above)
Comment (rigdomin-cisco) [reply discover/models/event_test.go]: done

File: discover/models/service.go (hunk 12 above)
Comment (rigdomin-cisco) [reply discover/models/service.go]: done

File: discover/models/service.go (hunk 13 above)
Comment (rigdomin-cisco) [reply discover/models/service.go]: done

File: discover/models/service.go (hunk 14 above)
Comment (rigdomin-cisco) [reply discover/models/service.go]: done

File: discover/models/service_test.go (hunk 15 above)
Comment (rigdomin-cisco) [reply discover/models/service_test.go]: done

File: discover/queue.go (hunk 16 above)
Comment (rigdomin-cisco) [reply discover/queue.go]: done

File: discover/queue_test.go (hunk 17 above)
Comment (rigdomin-cisco) [reply discover/queue_test.go]: done

File: discover/router.go (hunk 18 above)
Comment (rigdomin-cisco) [reply discover/router.go]: ticker set to updater

File: fixtures/rewrites.yaml (hunk 19 above)
Comment (rigdomin-cisco) [reply fixtures/rewrites.yaml]: done

File: go.mod (hunk 20 above)
Comment (rigdomin-cisco) [reply go.mod]: done

File: main.go (hunk 21 above)
Comment (rigdomin-cisco) [reply main.go]: done

File: .golangci.yml (hunk 22 above)
Comment (rigdomin-cisco) [reply .golangci.yml:397]: done

File: .golangci.yml [hunk 23]
```diff
@@ -343,6 +346,7 @@ linters:
     - nestif  # TODO(kenjones): refactor code
     - scopelint
     - stylecheck
+    - tagliatelle
```
Comment (kenjones-cisco) [inline .golangci.yml:349]: Why was this needed?

File: .golangci.yml [hunk 24]
```diff
@@ -379,6 +383,22 @@ issues:
...
+      linters:
+        - forbidigo
+
+    # TODO(rigdomin): handler this linter at a later time
+    - path: formatter/markdown.go|handlers/reference/reference.go
+      text: "reassigned, but never used afterwards"
+      linters:
+        - wastedassign
+
+    - path: spec/spec.go
```
Comment (kenjones-cisco) [inline .golangci.yml:403]: can you add a comment indicating how far outside of the configured bounds this is for a reference?

File: discover/controller.go [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+// example, an event for a service creation will see a service registry without
+// the service if the event is immediately followed by the service deletion
+// event.
+//
+// Handlers execute on the single worker queue in the order they are appended.
+// Handlers receive the notification event and the associated object.  Note
+// that all handlers must be appended before starting the watcher.
+type watcher interface {
+	// AppendServiceHandler notifies about changes to the service catalog.
+	AppendServiceHandler(f func(*models.Service, models.Event)) error
```
Comment (kenjones-cisco) [inline discover/controller.go]: since we never return an error, it would make sense to remove the return parameter

File: discover/controller.go [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+
+	// AppendDeploymentHandler notifies about change to the deployment catalog.
+	AppendDeploymentHandler(f func(*models.Deployment, models.Event)) error
```
Comment (kenjones-cisco) [inline discover/controller.go]: since we never return an error, it would make sense to remove the return parameter

File: discover/discover.go [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+func (d *Discoverer) Run() {
+	d.discover()
+
+	if d.services != nil {
+		go d.services.Run(d.stop)
+	}
+}
+
+// Specs returns discovered API specs.
+func (d *Discoverer) Specs() map[string][]byte {
```
Comment (kenjones-cisco) [inline discover/discover.go:81]: should probably include a mutex when accessing the specs, either reading or writing as there could end up being a high number of concurrent requests for the information.

File: discover/discover.go [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+
+func (d *Discoverer) updateServices(s *models.Service, e models.Event) {
+	log().Debugf("(Discover Handler) Service: %v Event: %v", s, e)
+
+	if sets.NewString(viper.GetStringSlice(config.DiscoveryServiceIgnoreList)...).Has(s.Hostname) {
+		log().Debugf("%v service is ignored", s.Hostname)
+
+		return
+	}
+
```
Comment (kenjones-cisco) [inline discover/discover.go:116]: it might be a good idea to leverage a mutex here to avoid multiple events triggering a discover call in parallel.

File: discover/discover.go [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+	case models.EventDelete:
+		d.data.services.Delete(s)
+	}
+
+	d.discover()
+}
+
+func (d *Discoverer) updateDeployments(dpl *models.Deployment, e models.Event) {
+	log().Debugf("(Discover Handler) Deployment: %v Event: %v", dpl, e)
+
```
Comment (kenjones-cisco) [inline discover/discover.go:129]: it might be a good idea to leverage a mutex here to avoid multiple events triggering a discover call in parallel.

File: discover/models/service.go [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+	switch p {
+	case ProtocolHTTP, ProtocolHTTP2, ProtocolGRPC:
+		return true
+	default:
+		return false
+	}
+}
+
+// serviceKey generates a service key for a collection of ports.
+func serviceKey(hostname string, servicePorts PortList) string {
```
Comment (kenjones-cisco) [inline discover/models/service.go]: where is this being used at?

File: discover/apispecs.go [hunk 31]
```diff
@@ -0,0 +1,226 @@
...
+			if !port.Protocol.IsHTTP() {
+				continue
+			}
+
+			hostName := service.Hostname
+			portNum := port.Port
+
+			svcSpec, err := loadSpec(fmt.Sprintf("%s:%d", hostName, portNum))
+			if err != nil {
+				log().Errorf("unable to load specs for %v: %v", hostName, err)
```
Comment (kenjones-cisco) [inline discover/apispecs.go]: `WithError(err)`

File: discover/apispecs.go [hunk 32]
```diff
@@ -0,0 +1,226 @@
...
+
+				continue
+			}
+
+			path, data, err := processSpec(hostName, rewritesDoc.Spec(), svcSpec)
+			if err != nil {
+				log().Errorf("unable to load specs for %v: %v", hostName, err)
```
Comment (kenjones-cisco) [inline discover/apispecs.go]: `WithError(err)`

File: discover/openapi.go [hunk 33]
```diff
@@ -0,0 +1,38 @@
+package discover
```
Comment (kenjones-cisco) [inline discover/openapi.go]: any reason this was placed here vs. within apispecs.go? given how closely related these are?

File: config/configurations.go [hunk 34]
```diff
@@ -36,7 +36,24 @@ const (
 	SpecFilename    = "spec-filename"
```
Comment (kenjones-cisco) [inline config/configurations.go:36]: all of the new configurations are actively being used within the code base, correct?

File: discover/apispecs.go [hunk 35]
```diff
@@ -0,0 +1,226 @@
...
+)
+
+const (
+	extKeyVisibility = "x-visibility"
+	extKeyGroupBy    = "x-groupby"
+
+	groupByDefault = "Common APIs"
+)
+
+func (d *Discoverer) fetchAPISpecs() map[string][]byte {
```
Comment (kenjones-cisco) [inline discover/apispecs.go:25]: it would help to refactor this, such that fetching the rewritesDoc is a function
looping through the services and ignoring service items that are not needed stays here, but the actual loading and processing happens separately to keep this overall logic simpler.
See mesh-discovery for hints.

File: .golangci.yml (hunk 23 above)
Comment (rigdomin-cisco) [reply .golangci.yml:349]: image

File: .golangci.yml (hunk 24 above)
Comment (rigdomin-cisco) [reply .golangci.yml:403]: commented

File: discover/controller.go (hunk 25 above)
Comment (rigdomin-cisco) [reply discover/controller.go]: done

File: discover/apispecs.go (hunk 31 above)
Comment (rigdomin-cisco) [reply discover/apispecs.go]: done

File: config/configurations.go (hunk 34 above)
Comment (rigdomin-cisco) [reply config/configurations.go:36]: removed unused configs

File: discover/openapi.go (hunk 33 above)
Comment (rigdomin-cisco) [reply discover/openapi.go]: This was result of modeling after existing implementations in IAM.
After trimming all the excess code, it does make sense to move this over.

Moved into `apispecs.go`

File: discover/apispecs.go (hunk 32 above)
Comment (rigdomin-cisco) [reply discover/apispecs.go]: done

File: discover/models/service.go (hunk 30 above)
Comment (rigdomin-cisco) [reply discover/models/service.go]: not used in any real code, was leftover from mesh code.
removed.

File: discover/controller.go (hunk 26 above)
Comment (rigdomin-cisco) [reply discover/controller.go]: removed

File: discover/discover.go (hunk 27 above)
Comment (rigdomin-cisco) [reply discover/discover.go:81]: done

File: discover/discover.go (hunk 28 above)
Comment (rigdomin-cisco) [reply discover/discover.go:116]: done

File: discover/discover.go (hunk 29 above)
Comment (rigdomin-cisco) [reply discover/discover.go:129]: done

File: discover/apispecs.go (hunk 35 above)
Comment (rigdomin-cisco) [reply discover/apispecs.go:25]: done

File: discover/discover.go [hunk 36]
```diff
@@ -0,0 +1,152 @@
...
+	dm := models.NewDeploymentMap()
+
+	d := &Discoverer{
+		services: ctlg,
+		data: &state{
+			services:    &sm,
+			deployments: &dm,
+		},
+		stop:  make(chan struct{}),
+		vpr:   vprcpy,
```
Comment (kenjones-cisco) [inline discover/discover.go]: is `vpr` being used anywhere?

File: discover/discover.go [hunk 37]
```diff
@@ -0,0 +1,152 @@
...
+	//  - rewrite spec details for Schema, Security Definitions, Security
+	specs := d.fetchAPISpecs()
+	if len(specs) == 0 {
+		return
+	}
+
+	log().Debug("successfully processed API changes")
+
+	// update local cache with latest service specs
+	d.sLock.Lock()
```
Comment (kenjones-cisco) [inline discover/discover.go]: the locking should be done before the call to `fetchAPISpecs`

File: discover/discover.go (hunk 36 above)
Comment (rigdomin-cisco) [reply discover/discover.go:62]: Should've been used inside Discover's private methods.

Comment (rigdomin-cisco) [reply discover/discover.go]: removed

File: discover/discover.go (hunk 37 above)
Comment (rigdomin-cisco) [reply discover/discover.go]: done

File: discover/discover.go [hunk 38]
```diff
@@ -0,0 +1,145 @@
...
+func (d *Discoverer) updateServices(s *models.Service, e models.Event) {
+	log().Debugf("(Discover Handler) Service: %v Event: %v", s, e)
+
+	if sets.NewString(viper.GetStringSlice(config.DiscoveryServiceIgnoreList)...).Has(s.Hostname) {
+		log().Debugf("%v service is ignored", s.Hostname)
+
+		return
+	}
+
+	d.eLock.Lock()
```
Comment (kenjones-cisco) [inline discover/discover.go]: since there is a lock on `discover()` now for making sure the writes happen properly. The 2nd lock could actually result in things being unnecessarily slowed down.

**Comment by kenjones-cisco** [comment]
Have you tried testing this out locally?

File: discover/discover.go (hunk 38 above)
Comment (rigdomin-cisco) [reply discover/discover.go]: sure. will remove.

**Comment by rigdomin-cisco** [comment]
> Have you tried testing this out locally?

No. I need to test this in a live local environment.

- [x] Test Locally

File: Dockerfile.dev [hunk 39]
```diff
@@ -15,7 +15,7 @@ RUN wget -O /usr/local/bin/jq "https://github.com/stedolan/jq/releases/download/
 
 RUN go get golang.org/x/tools/cmd/goimports \
     && go get github.com/mitchellh/gox \
-    && wget -O - -q https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh| sh -s -- -b /usr/local/bin \
+    && wget -O - -q https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh| sh -s -- -b /usr/local/bin v1.41.1 \
```
Comment (rigdomin-cisco) [inline Dockerfile.dev:18]: Aligning golangci version to better match our usage

File: .golangci.yml [hunk 40]
```diff
@@ -388,6 +388,11 @@ issues:
       linters:
         - forbidigo
 
+    - path: discover/apispecs_test.go
+      text: "Possible resource leak"
```
Comment (rigdomin-cisco) [inline .golangci.yml:392]: This was caused by a UT, where I create and defer a GenServer() only needed for UTs.
The server instance is spun up and brought for each iteration of table tests.

**Comment by rigdomin-cisco** [comment]
Will Squash Later

- [x] Squash Commits

**Comment by rigdomin-cisco** [comment]
Tested and currently running for review. 
See DOCS PR.

**Comment by rigdomin-cisco** [comment]
Ready for final review

**Comment by kenjones-cisco** [comment]
@rigdomin-cisco  Squash and then we can merge

**Comment by kenjones-cisco** [comment]
Thanks Rigo!
